jax/interpreters/pxla.py

```diff
@@ -67,32 +161,14 @@ def shard_args(devices, assignments, axis_size, args):
   for a, arg in enumerate(args):
     # The shard_arg_handlers allow an extensible set of types to be sharded, but
     # inline handling for ShardedDeviceArray as a special case for performance
-    if type(arg) is ShardedDeviceArray:
-      if nrep == len(arg.device_buffers):
-        # The argument is already prepared for the right number of replicas, so
-        # we just ensure that buf[r] is on devices[r] for each replica index r
-        # TODO(mattjj): compared to the other case, this logic has less looping
-        # but could incur more device-to-device data movement
-        for r, buf in enumerate(arg.device_buffers):
-          buffers[r][a] = buf if buf.device() == devices[r] else buf.copy_to_device(devices[r])
-      else:
-        # The argument is prepared for a different number of replicas, so for
```

Comment:
This old logic was doing a shuffle, right? This is not done anymore in the new logic. Was it an unnecessary optimization?

Reply:
Not sure if you saw, but there's a new version of this logic in the new `_shard_sharded_device_array_slow_path` method, which we call via `shard_arg_handlers`. That said, I'm not sure anyone depends on this logic, and we didn't have any test coverage of it until this PR.